### ISSUE TITLE
[clang] why is clang not support -Wmaybe-uninitialized

### ISSUE THREAD

**Issue by CoTinker** [llvm/llvm-project#72481]
demo.c:

```
#include <stdio.h>
void foo (int a) {
   printf("%d\n",a);
}
int main() {
  int x;
  int y;
  scanf("%d",&y);
  switch (y)
    {
    case 1: x = 1;
      break;
    case 2: x = 4;
      break;
    case 3: x = 5;
    }
  foo (x);
}
```
compile command

`gcc demo.c -o demo -O1 -Wall`

report warning:
```
demo.c: In function 'main':
demo.c:17:3: warning: 'x' may be used uninitialized in this function [-Wmaybe-uninitialized]
   foo (x);
   ^~~~~~~
```
https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wmaybe-uninitialized
but clang not support this warning.
I wonder why the llvm community does not support this alarm and whether it will be supported  in the future.

**Comment by CoTinker** [comment]
> One very annoying aspect of gcc's -Wmaybe-uninitialized is its many false positives. This is the reason that many projects disable the warning, because working around such false positives is very counterproductive. Typically it is easier to run a "real" static analysis tool, such as scan-build.
I see, thanks, so it means that the alarm item is actually defective?

**Comment by CoTinker** [comment]
> One very annoying aspect of gcc's -Wmaybe-uninitialized is its many false positives. This is the reason that many projects disable the warning, because working around such false positives is very counterproductive. Typically it is easier to run a "real" static analysis tool, such as scan-build.

I see, thanks, so it means that the alarm item is actually defective?

**Comment by DimitryAndric** [comment]
> > One very annoying aspect of gcc's -Wmaybe-uninitialized is its many false positives. This is the reason that many projects disable the warning, because working around such false positives is very counterproductive. Typically it is easier to run a "real" static analysis tool, such as scan-build.
> 
> I see, thanks, so it means that the alarm item is actually defective?

Compiler warnings are always an advice, not a hard error. This is because for almost all warning types, there exist situations where the warning is completely safe to ignore. (In such cases the code can be slightly adjusted to work around the warning, or pragmas can be inserted to disable the warning temporarily.)

In the particular case of -Wmaybe-uninitialized, the documentation (https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wmaybe-uninitialized) states:

> if there exists a path from the function entry to a use of the object that is initialized, but there exist some other paths for which the object is not initialized, the compiler emits a warning if it cannot prove the uninitialized paths are not executed at run time.

and:

> These warnings are made optional because GCC may not be able to determine when the code is correct in spite of appearing to have an error.

It's a bit of a pity that gcc enables this warning with -Wall, since many projects use -Wall.

**Comment by nathanchance** [comment]
> but clang not support this warning.
> I wonder why the llvm community does not support this alarm and whether it will be supported in the future.

`clang`'s version is `-Wconditional-uninitialized`: https://godbolt.org/z/oW3sTWErz

While I agree that `-Wmaybe-uninitialized` has many pitfalls (which is why the Linux kernel has disabled it under the "normal" build options), I am personally surprised that `-Wsometimes-uninitialized` (a less wishy-washy version of `-Wconditional-uninitialized`) does not fire for this because if none of the cases in the switch statement are taken, `x` is definitely uninitialized (rather than "maybe").

**Comment by CoTinker** [comment]
> > but clang not support this warning.
> > I wonder why the llvm community does not support this alarm and whether it will be supported in the future.
> 
> `clang`'s version is `-Wconditional-uninitialized`: https://godbolt.org/z/oW3sTWErz
> 
> While I agree that `-Wmaybe-uninitialized` has many pitfalls (which is why the Linux kernel has disabled it under the "normal" build options), I am personally surprised that `-Wsometimes-uninitialized` (a less wishy-washy version of `-Wconditional-uninitialized`) does not fire for this because if none of the cases in the switch statement are taken, `x` is definitely uninitialized (rather than "maybe").

It was very helpful, thank you very much.

**Comment by dwblaikie** [comment]
>  I am personally surprised that -Wsometimes-uninitialized (a less wishy-washy version of -Wconditional-uninitialized) does not fire for this
`-Wsometimes-uninitialized` only catches the cases that aren't already caught by `conditional-unitialized`

(essentially `-Wsometimes-uninitialized` is "higher false positive cases" that don't qualify for `-Wconditional-uninitialized` - it's intended to be a simpler model that the diagnostic is only under one group, not multilpe)

**Comment by nathanchance** [comment]
> `-Wsometimes-uninitialized` only catches the cases that aren't already caught by `conditional-unitialized`
> 
> (essentially `-Wsometimes-uninitialized` is "higher false positive cases" that don't qualify for `-Wconditional-uninitialized` - it's intended to be a simpler model that the diagnostic is only under one group, not multilpe)

Is this explanation backwards? I thought `-Wsometimes-uninitialized` was supposed to have lower false positive cases than `-Wconditional-uninitialized` (at least that appears to be the case based on [all the fixes that have been done to the Linux kernel for these warnings](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/?qt=grep&q=-Wsometimes-uninitialized)) and based on ba8071ec81eb5c39f8ebb869a18181a1ed25c782, it sounds like `-Wconditional-uninitialized` is the catch-all, not `-Wsometimes-uninitialized`.

**Comment by dwblaikie** [comment]
> > `-Wsometimes-uninitialized` only catches the cases that aren't already caught by `conditional-unitialized`
> > (essentially `-Wsometimes-uninitialized` is "higher false positive cases" that don't qualify for `-Wconditional-uninitialized` - it's intended to be a simpler model that the diagnostic is only under one group, not multilpe)
> 
> Is this explanation backwards? I thought `-Wsometimes-uninitialized` was supposed to have lower false positive cases than `-Wconditional-uninitialized` (at least that appears to be the case based on [all the fixes that have been done to the Linux kernel for these warnings](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/log/?qt=grep&q=-Wsometimes-uninitialized)) and based on [ba8071e](https://github.com/llvm/llvm-project/commit/ba8071ec81eb5c39f8ebb869a18181a1ed25c782), it sounds like `-Wconditional-uninitialized` is the catch-all, not `-Wsometimes-uninitialized`.

Right, yes, I mispoke/had that around backwards: https://godbolt.org/z/nxj6GTG5h
